sdk/src/trace/probability_sampler.cc

```diff
@@ -0,0 +1,127 @@
+// Copyright 2020, Open Telemetry Authors
+// Copyright 2017, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "opentelemetry/sdk/trace/probability_sampler.h"
+
+#include <cmath>
+#include <cstdint>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+namespace trace_api = opentelemetry::trace;
+
+ProbabilitySampler::ProbabilitySampler(
+  double probability, bool defer_parent, SamplingBehavior sampling_behavior 
+)
+: threshold_(CalculateThreshold(probability)), probability_(probability),
+  defer_parent_(defer_parent), sampling_behavior_(sampling_behavior) {}
+
+SamplingResult ProbabilitySampler::ShouldSample(
+  const SpanContext *parent_context,
+  trace_api::TraceId trace_id,
+  nostd::string_view name,
+  trace_api::SpanKind span_kind,
+  const trace_api::KeyValueIterable &attributes) noexcept
+{
+  // Check if current Span should be sampled with current Sampling Behavior
+  if (sampling_behavior_ == SamplingBehavior::DETACHED_SPANS_ONLY
+      && parent_context != nullptr && !parent_context->is_remote) {
+    return { Decision::NOT_RECORD, nullptr };
+  }
+
+  if (sampling_behavior_ == SamplingBehavior::ROOT_SPANS_ONLY
+      && parent_context != nullptr) {
+    return { Decision::NOT_RECORD, nullptr };
+  }
+
+  // Check if sampling decision should be defered to parent span, if applicable 
+	if (parent_context == nullptr || !defer_parent_)
+  {
+    if (threshold_ == 0) return { Decision::NOT_RECORD, nullptr };
```

Comment:
Consider pulling this check up farther, since 0 threshold always returns NOT_RECORD.

Reply:
If the behavior is to defer to the parent, it won't necessarily return NOT_RECORD in that case. The threshold only comes into effect once it's determined that either A) there is no parent span, or B) the sampler is configured to not defer to it's parent span.